activesupport/lib/active_support/hash_with_indifferent_access.rb

```diff
@@ -94,7 +94,11 @@ def self.[](*args)
     #
     # This value can be later fetched using either +:key+ or +'key'+.
     def []=(key, value)
-      regular_writer(convert_key(key), convert_value(value, for: :assignment))
+      if value.frozen?
+        regular_writer(convert_key(key), convert_value(value))
+      else
+        regular_writer(convert_key(key), convert_value(value, for: :assignment))
+      end
```

Comment:
Sorry, but I think this fix is at the wrong level, because `convert_value` recurses into `Hash` and `Array` values, and may hit a frozen object deeper down. Probably better for `convert_value` to check `frozen?` internally.